docs/pages/access-controls/reference.mdx

```diff
@@ -418,14 +320,253 @@ spec:
+</Admonition>
+
+Teleport roles support matching resource labels with predicate expressions when
+you need to:
+
+- combine logic with OR and AND operators
+- perform matching on label keys
+- implement negative matches
+
+The following example role would allow access to any nodes labeled `env=staging`
+or labeled `team=<team>`, where `<team>` is one of the values of the user's
+`teams` trait:
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: example-role
+spec:
+  allow:
+    node_labels_expression: |
+      labels["env"] == "staging" ||
+      contains(user.spec.traits["teams"], labels["team"])
+```
+
+Typically you will only want to use one of `<kind>_labels` or
+`<kind>_labels_expression` in a single role, but you are allowed to specify
+both.
+If you have both in a deny rule, the matching is greedy, if either one matches
+access will be denied.
+In an allow rule, the matching is not greedy, if both are set they both have to
+match for access to be allowed.
+
+#### Fields
+
+The `<kind>_labels_expression` fields have the same purpose as the matching
+`<kind>_labels` fields, but support predicate expressions instead of label
+matchers.  They can be used in the following fields of role `spec.allow` and
+`spec.deny` conditions:
+
+- `node_labels_expression`
+- `app_labels_expression`
+- `cluster_labels_expression`
+- `kubernetes_labels_expression`
+- `db_labels_expression`
+- `db_service_labels_expression`
+- `windows_desktop_labels_expression`
+- `group_labels_expression`
+
+#### Variables
+
+You can use the following template variables in label expression fields:
+
+|Variable|Type|Description|
+|---|---|---|
+|`{{user.spec.traits}}`|`map[string][]string`|All of the user's traits. Use `{{user.spec.traits<key>}}` to get a list of trait values for the specified key.|
+|`{{labels}}`|`map[string]string`|All of the user's labels. Use `{{labels.<key>}}` to get the value for the specified key.|
+
+#### Functions
+
+You can use the following functions in label expression fields:
+
+|Function|Description|
+|---|---|
+|`{{contains(list,item)}}`|Returns `true` if the list of strings in `list` contains the specified `item` and `false` otherwise.|
+|`{{regexp.match(list,regexp)}}`|Returns `true` if the regular expression in `regexp` matches any of the strings in list of strings `list`. Returns `false` otherwise.|
+|`{{regexp.replace(list,regexp,replacement)}}`|Checks the list of strings `list`, returning a new list where each item that matches `regexp` is replaced with `replacement`. You can use `$n` in the replacement string to indicate a submatch group, e.g., `$1` to indicate the first submatch.|
+|`{{strings.upper(list)}}`|Returns the new list of strings that results from uppercasing each item in `list`, a list of strings.|
+|`{{strings.lower(list)}}`|Returns the new list of strings that results from lowercasing each item in `list`, a list of strings.|
+
+### Predicates for access to infrastructure resources
```

Comment:
```suggestion
### Predicates for access to infrastructure resources
```
Might just be my ignorance, but I don't see the connection between this heading and the paragraph and bullets that follow.

Reply:
I don't think of the below as predicates. I'll borrow a wikipedia definition for predicate:
> A predicate is a statement or mathematical assertion that contains variables, sometimes referred to as predicate variables, and may be true or false depending on those variables’ value or values.

Label expressions and search filters *are* predicates, because they evaluate to true or false. The below like `logins: "{{internal.logins}}"` is not a predicate that is true or false. It is a template that can be populated with the user traits to produce a value, like `logins: ["nic", "root"]`